File: scripts/gc_translation/prompt.py
"""Assemble the translation system prompt."""
from __future__ import annotations

from scripts.gc_translation.glossary import format_glossary_for_prompt

TERMINOLOGY_RULES = """\
MANDATORY Vietnamese terminology rules (follow without exception):

1. Use "Đức Chúa Giê-su" instead of "Chúa Giê-su", "Jesus", or "Giê-xu".
2. Chúa / God "ban phước" (blesses with authority). People "chúc phước"
   (wish blessings). NEVER use "chúc phước" when the subject is God.
3. Use "Sa-bát" (not "Sabát").
4. Use "Do Thái Giáo" (not "Giu-đa-izt").
5. Use "Cơ-đốc" (not "Cơ Đốc").
6. Capitalize divine names correctly: "Đức Chúa Trời", "Đức Thánh Linh",
   "Kinh Thánh", "Đức Chúa Giê-su".
"""

BIBLE_RULE = """\
Wherever the English text quotes a Bible verse directly (with an inline or
parenthetical reference like "(Luke 21:20)" or "Matthew 24:20 says..."),
replace the quoted English verse text with a sentinel of the form:

    [[BIBLE:<Book> <Chapter>:<Verse>]]
    [[BIBLE:<Book> <Chapter>:<VerseStart>-<VerseEnd>]]  (for ranges)

Use the English book name exactly (Matthew, Luke, Genesis, 1 Corinthians...).
Do NOT translate the quoted verse text — emit only the sentinel. The reference
text itself ("Luke 21:20") in the surrounding prose should be translated
normally using the glossary's book-name renderings.
"""

OUTPUT_RULE = """\
Output ONLY the Vietnamese translation as pure Markdown. Preserve paragraph
breaks, h2 headings (## …), and emphasis. No preamble, no explanation, no
surrounding commentary. Do not wrap the output in code fences.
"""

ROLE = """\
You are an expert Vietnamese translator specializing in Christian theological
texts for a Seventh-day Adventist audience. Your register is reverent, clear,
and consistent with the 1934-era Vietnamese Protestant tradition.
"""


def build_system_prompt(glossary: dict[str, str]) -> str:
    glossary_block = format_glossary_for_prompt(glossary) if glossary else ""
    return "\n\n".join([
        ROLE.strip(),
        TERMINOLOGY_RULES.strip(),
        "GLOSSARY — use these Vietnamese renderings verbatim for the English terms on the left:",
        glossary_block,
        BIBLE_RULE.strip(),
        OUTPUT_RULE.strip(),
    ])

File: scripts/egw_translate/lib/prompt.ts
import type { Glossary } from "./glossary";
import { formatGlossaryTable } from "./glossary";

export interface PromptInputs {
  glossary: Glossary;
  paragraphCitationPrefix: string; // e.g., "COL"
}

/**
 * Build the system prompt used for every translation chunk. The prompt is stable within a
 * run, so a future migration to the Anthropic SDK can prompt-cache it for cost savings.
 */
export function buildSystemPrompt({ glossary, paragraphCitationPrefix }: PromptInputs): string {
  const sections: string[] = [];

  sections.push(
    "You are an expert Vietnamese translator for Seventh-day Adventist literature by Ellen G. White. " +
    "Translate the provided English passage into natural, reverent Vietnamese.",
  );

  sections.push(
    "# Terminology rules (HARD — never deviate)\n" +
    "- Use \"Đức Chúa Giê-su\" (never \"Chúa Giê-su\", \"Jesus\", or \"Giê-xu\").\n" +
    "- Use \"ban phước\" when God is the subject; \"chúc phước\" only when a human is the subject.\n" +
    "- Use \"Sa-bát\" (never \"Sabát\").\n" +
    "- Use \"Do Thái Giáo\" (never \"Giu-đa-izt\").\n" +
    "- Use \"Cơ-đốc\" (never \"Cơ Đốc\").\n" +
    "- Capitalize divine names correctly: \"Đức Chúa Trời\", \"Đức Thánh Linh\", \"Kinh Thánh\", \"Đức Chúa Giê-su\".",
  );

  sections.push(
    "# Glossary (MANDATORY — use these Vietnamese renderings verbatim when the English term appears)\n" +
    formatGlossaryTable(glossary),
  );

  sections.push(
    "# Bible quotations\n" +
    "When the passage quotes or paraphrases a Bible verse with an explicit reference (e.g., \"Matthew 13:31\", \"Luke 15:1-4\"), " +
    "REPLACE the quoted English verse text plus its reference with a single sentinel of the form `[[BIBLE:Book Chapter:Verse]]` " +
    "(or `[[BIBLE:Book Chapter:Verse-Verse]]` for ranges, or `[[BIBLE:Book Chapter:V1,V2]]` for comma lists). " +
    "Do NOT translate the quoted verse text yourself — a post-processor will substitute the canonical Vietnamese (VI1934) translation. " +
    "Use the CANONICAL English book name exactly as given in the English source.",
  );

  sections.push(
    "# Paragraph citations\n" +
    `Each English paragraph ends with a citation like \`${paragraphCitationPrefix} N.M\` (e.g., \`${paragraphCitationPrefix} 17.1\`). ` +
    "Preserve these citations VERBATIM at the end of the corresponding Vietnamese paragraph.",
  );

  sections.push(
    "# Output format\n" +
    "- Output Vietnamese Markdown ONLY. No preamble, no explanation, no wrapping in code fences.\n" +
    "- Preserve `## ` section headings (translate the heading text into Vietnamese).\n" +
    "- Separate paragraphs with a single blank line.\n" +
    "- Do not add content that was not in the source.",
  );

  return sections.join("\n\n");
}
